uobtheatre/payments/payment_methods.py

```diff
@@ -0,0 +1,316 @@
+        """Send payment to point of sale device.
+
+        Parameters:
+            value (int): Amount of money being payed
+            pay_object (Payable): The object being payed for
+
+        Raises:
+            SquareException: If the request was unsuccessful.
+        """
+        body = {
+            "idempotency_key": str(uuid.uuid4()),
+            "checkout": {
+                "amount_money": {
+                    "amount": value,
+                    "currency": "GBP",
+                },
+                "reference_id": pay_object.payment_reference_id,
+                "device_options": {
+                    "device_id": self.device_id,
+                },
+            },
+        }
+        response = self.client.terminal.create_terminal_checkout(body)
+        if not response.is_success():
+            raise SquareException(response)
+
+    @classmethod
+    def handle_webhook(cls, data: dict, signature: str):
+        """Handle checkout event for terminal api.
+
+        Args:
+            data (dict): The data provided in the webhook event
+            signature (str): The signature of the requeset
+
+        Raises:
+            ValueError: If the signature is invalid
+        """
+        if not cls.is_valid_callback(data, signature):
+            raise ValueError("Invalid signature")
+
+        if data["type"] == "terminal.checkout.updated":
+            cls.handle_terminal_checkout_updated_webhook(data["data"])
+
+    @classmethod
+    def handle_terminal_checkout_updated_webhook(cls, data: dict):
+        """Handle a terminal checkout update event
+
+        Function to handle a square webhook containing terminal checkout
+        udpate. If the status of the checkout is completed the booking can be
+        completed and a payment object can be created.
+        """
+
+        from uobtheatre.bookings.models import Booking
+
+        checkout = data["object"]["checkout"]
+        booking = Booking.objects.get(reference=checkout["reference_id"])
+
+        if checkout["status"] == "COMPLETED":
+            cls.create_payment_object(
+                booking,
+                checkout["amount_money"]["amount"],
+                provider_payment_id=checkout["payment_ids"][0],
+                currency=checkout["amount_money"]["currency"],
+            )
+            booking.complete()
+
+    @classmethod
+    def list_devices(cls) -> list[dict]:
+        """List the device codes available on square.
+
+        Returns:
+            list of dict: A list of dictionaries which store the device code
+                info. When connection to a device the `device_id` should be
+                used.
+
+        Raises:
+            SquareException: If the square request returns an error
+        """
+        response = cls.client.devices.list_device_codes()
+        if not response.is_success():
+            raise SquareException(response)
+        return response.body["device_codes"]
```

Comment:
Does this only list device codes that have been "paired"? Also must have the product code of `TERMINAL_API` or something

Reply:
I've added some filters for this, see this test for an example
https://github.com/BristolSTA/uobtheatre-api/blob/0528b39f108f07dc80f48ac5336b7bb45a71475f/uobtheatre/payments/test/test_schema.py#L214

Reply:
Yes, but I'm wondering if we actually want to expose all these? Other device codes are only used for the POS app (I think). I don't think the front end should be concerned about having to pass TERMINAL API, I think the method should only return terminal api because that basically is/are our devices? I can't think of a situation where we might want non terminal api devices (at least retrieval - we might want ability to create a device code for POS sign on later for puck)

Reply:
I think the point about `TERMINAL API` is valid, however, I think it's likely in the future we have greater integration with square devices so returning unpaired and nonterminal devices will become something we want to do. I've updated the filters to filter based on the payment method as this will already be required by the front end to make the pos payment. This should handle the setting of `TERMINAL_API`. You then have a boolean `paired.